course3/homework-list/list.h

```diff
@@ -0,0 +1,105 @@
+#pragma once
+
+#include <memory>
+
+
+namespace Karun
+{
+
+template<typename T>
+class List
+{
+public:
+	class Node
+	{
+	friend class List;
+	
+	public:
+		Node(const T& data, std::unique_ptr<Node>& next)
+		: 	mData(data)
+		{
+			mNext = std::move(next);
+		}
+		
+	private:
+		T mData;
+		std::unique_ptr<Node> mNext;
```

Comment:
I don't know how about you, but I find this much cleaner now: Node _owns_ mNext (hence the smart pointer), while iterator only passively points at it (hence the raw pointer).